File: CSC310-Python/Module12-Spanning-Trees/debug01/kruskal.py
"""Kruskal's minimum spanning tree algorithm (CLRS Section 21.2).

kruskal(vertices, edges) sorts the edges by weight and accepts each edge
whose endpoints currently lie in different components, tracking the
components with a disjoint-set forest (union by rank, path compression).
The main block runs the algorithm on the nine-vertex example graph of CLRS Chapter 21
and reports the chosen edges, their count, and the total weight.
"""


class DisjointSet:
    def __init__(self, items):
        self.parent = {x: x for x in items}
        self.rank = {x: 0 for x in items}

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rx, ry = self.find(x), self.find(y)
        if rx == ry:
            return
        if self.rank[rx] < self.rank[ry]:
            rx, ry = ry, rx
        self.parent[ry] = rx
        if self.rank[rx] == self.rank[ry]:
            self.rank[rx] += 1


def kruskal(vertices, edges):
    ds = DisjointSet(vertices)
    tree = []
    # Python's sort is stable, so equal weights keep their listed order.
    for u, v, w in sorted(edges, key=lambda e: e[2]):
        if ds.parent[u] != ds.parent[v]:
            tree.append((u, v, w))
            ds.union(u, v)
    return tree


def main():
    # The nine-vertex example graph of CLRS Chapter 21
    vertices = list("abcdefghi")
    edges = [
        ("a", "b", 4), ("a", "h", 8), ("b", "c", 8), ("b", "h", 11),
        ("c", "d", 7), ("c", "f", 4), ("c", "i", 2), ("d", "e", 9),
        ("d", "f", 14), ("e", "f", 10), ("f", "g", 2), ("g", "h", 1),
        ("g", "i", 6), ("h", "i", 7),
    ]
    tree = kruskal(vertices, edges)
    print("edges chosen:")
    for u, v, w in tree:
        print(f"  {u}-{v} ({w})")
    print(f"tree edges: {len(tree)} (a spanning tree on {len(vertices)} "
          f"vertices has {len(vertices) - 1})")
    print("total weight:", sum(w for _, _, w in tree))


if __name__ == "__main__":
    main()

File: CSC310-Cpp/Module12-Spanning-Trees/debug01/kruskal.cpp
// kruskal.cpp
// Kruskal's minimum spanning tree algorithm (CLRS 21.2) with a disjoint-set
// forest. Edges are examined in order of increasing weight and an edge is
// added to the tree when its endpoints are in different components.
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

struct Edge {
    int u, v, weight;
};

int find(std::vector<int>& parent, int x) {
    if (parent[x] != x) {
        parent[x] = find(parent, parent[x]);
    }
    return parent[x];
}

void unite(std::vector<int>& parent, int a, int b) {
    parent[find(parent, a)] = find(parent, b);
}

std::vector<Edge> kruskal(int n, std::vector<Edge> edges) {
    std::sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b) {
        if (a.weight != b.weight) return a.weight < b.weight;
        if (a.u != b.u) return a.u < b.u;
        return a.v < b.v;
    });
    std::vector<int> parent(n);
    for (int i = 0; i < n; ++i) {
        parent[i] = i;
    }
    std::vector<Edge> tree;
    for (const Edge& e : edges) {
        if (parent[e.u] != parent[e.v]) {
            tree.push_back(e);
            unite(parent, e.u, e.v);
        }
    }
    return tree;
}

int main() {
    const std::string names = "abcdefghi";
    const int n = 9;
    std::vector<Edge> edges = {
        {0, 1, 4}, {0, 7, 8}, {1, 2, 8}, {1, 7, 11}, {2, 3, 7}, {2, 5, 4}, {2, 8, 2},
        {3, 4, 9}, {3, 5, 14}, {4, 5, 10}, {5, 6, 2}, {6, 7, 1}, {6, 8, 6}, {7, 8, 7}};

    std::vector<Edge> tree = kruskal(n, edges);
    int total = 0;
    std::cout << "edges chosen:\n";
    for (const Edge& e : tree) {
        std::cout << "  " << names[e.u] << "-" << names[e.v] << " (" << e.weight << ")\n";
        total += e.weight;
    }
    std::cout << "tree edges: " << tree.size() << " (a spanning tree on " << n
              << " vertices has " << n - 1 << ")\n";
    std::cout << "total weight: " << total << '\n';
    return 0;
}
